### ISSUE TITLE
Type error differences between `tsc` and `microbundle`

### ISSUE THREAD

**Issue by tom-sherman** [developit/microbundle#675]
Reproduction:

- Clone https://github.com/tom-sherman/immurl @ bb2644ecfe1bb4f2098c02df813acf5cf2d5de07
- `npm i`
- `npm run typecheck` - No type errors
- `npm run build` - Type errors from microbundle
- Uncomment methods from bottom of `lib/immurl.ts`
- `npm run typecheck` - Type errors from `tsc`
- `npm run build` - No type errors

Is this a bug or am I missing some configuration? It looks like microbundle is using different lib definitions to typescript here.

**Comment by marvinhagemeister** [comment]
@tom-sherman The repo link leads to a 404 for me. Is this a private repo by chance?

**Comment by tom-sherman** [comment]
@marvinhagemeister Whoops! Should be public now

**Comment by tom-sherman** [comment]
Hmmm, I have found the fix to be to set the "target" in tsconfig.json to "ES6" or above as mentioned [here](https://github.com/Microsoft/TypeScript/issues/23174).

Admitidly this was a mistake on my part, I don't want to target ES5 or below but the error suggests that microbundle isn't respecting my tsconfig?

**Comment by marvinhagemeister** [comment]
Microbundle overwrites both the `module` and `target` compiler options. Those two are the only properties that can't be overwritten.

This is done because we pass the transpiled files back through babel again which does all the downtranspilation for us. Setting it to something lower breaks `import()` syntax. And then we get loads of bugs reports again that `microbundle` would violate it's own no config setup :shrug: 

The default target used by `tsc` when invoked without configuration is `ES5`. There the `URLSearchParam` interface needs less properties to align with the spec.

So there are 2 solutions to this that I can think of:

1) Implement the full interface, instead of coding against the ES5 definition of `URLSearchParams`
2) Specify the lib `.d.ts` files manually (not recommended)

In conclusion 1) is the proper approach that ensures that your code is future proof. We highly recommend going with that :+1:

**Comment by tom-sherman** [comment]
@marvinhagemeister 

> Microbundle overwrites both the module and target compiler options. Those two are the only properties that can't be overwritten.

This is the part I was missing and isn't mentioned in the docs and I don't think many people would say that was obvious. Would you up for reviewing a PR that clarifies this in the README and strongly recommends to match whatever microbundle's target/module is in the tsconfig?

**Comment by marvinhagemeister** [comment]
Sure, mentioning it in the README is a good idea! We're always using the `esnext` target and `ESNext` as the module type. This ensures that `microbundle` works with any future TypeScript version. Happy to review any PRs :+1:
